lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1644,12 +1671,19 @@ bool LinearMapInfo::shouldDifferentiateInstruction(SILInstruction *inst) {
       [&](SILValue val) { return activityInfo.isActive(val, indices); });
   if (hasActiveOperands && hasActiveResults)
     return true;
-  // A `store`-like instruction does not have an SSA result, but has two
+  // `store`-like instructions do not have an SSA result, but have two
   // operands that represent the source and the destination. We treat them as
   // the input and the output, respectively.
-#define CHECK_INST_TYPE_ACTIVE_DEST(INST) \
-  if (auto *castInst = dyn_cast<INST##Inst>(inst)) \
-    return activityInfo.isActive(castInst->getDest(), indices);
+  // For `store`-like instructions whose destination is an element address from
```

Comment:
(Can be revisited later if you'd rather submit this improvement now instead of spending more time.)

Alternative to this new activity logic that might be less code and more robust: Modify variedness propagation to propagate variedness from the address to the array, and modify usefulness propagation to propagate usefulness from the array to the address.

This will make it robust to the thing about tuples that I commented about above. (Which might not actually be a problem as noted by @rxwei, but it still seems like it would be good to propagate activity though these things in case there are situations we haven't noticed.)

This will also make the addresses and projections show up as active in the debugging output, which will make it easier to understand. (I assume that they don't show up as active now.)

Do you think this approach is good and will work?

Reply:
Thanks for the suggestion!

Element addresses (`pointer_to_address` and `index_addr`) of active `array.uninitialized_intrinsic` applications are now marked active in https://github.com/apple/swift/pull/28320/commits/9558aa2a7ed9cc3c83d7b5f4531f3bfbd482f6e8. This makes sense semantically and simplifies `LinearMapInfo::shouldDifferentiateInstruction`.